xarray/backends/common.py

```diff
@@ -223,8 +223,13 @@ def set_variables(self, variables, check_encoding_set,
         for vn, v in iteritems(variables):
             name = _encode_variable_name(vn)
             check = vn in check_encoding_set
-            target, source = self.prepare_variable(
-                name, v, check, unlimited_dims=unlimited_dims)
+            if (vn not in self.variables or
+                    (getattr(self, '_mode', False) != 'a')):
```

Comment:
I don't think we need the explicit check here for the mode here. With `mode='w'`, the file should already be starting from scratch.